### DISCUSSION TITLE
How to think about writing de/serializers in Fable?

### DISCUSSION THREAD

**Discussion by NickDarvey** [fable-compiler/Fable#2909] [Q&A] [answered]
I've written [a little thing](https://github.com/primacydotco/Symfig/blob/main/src/Config.fs) to help us with our environment configuration of our fsharp applications. It only works in .NET at the moment, but I want to use this in our browser applications with Fable too. 

It's reflection-heavy so it doesn't work with Fable. I'm guessing I need to take quite a different approach but I'm not sure where to start. Can anyone point me toward techniques I could be using to implement Fable-compatible `write` and `read` implementations?

```
.\src\Config.fs(66,5): (66,23) error FABLE: System.Type.IsAssignableTo is not supported by Fable
.\src\Config.fs(94,18): (94,40) error FABLE: System.Type.GetProperty is not supported by Fable - Inline call from .(192,4)
.\src\Config.fs(22,11): (22,17) error FABLE: Cannot get type info of generic parameter t. Fable erases generics at runtime, try inlining the functions so generics can be resolved at compile time.
.\src\Config.fs(28,11): (28,17) error FABLE: Cannot get type info of generic parameter t. Fable erases generics at runtime, try inlining the functions so generics can be resolved at compile time.
.\src\Config.fs(105,7): (105,13) error FABLE: Cannot get type info of generic parameter a$. Fable erases generics at runtime, try inlining the functions so generics can be resolved at compile time.
.\src\Config.fs(148,7): (148,13) error FABLE: Cannot get type info of generic parameter config. Fable erases generics at runtime, try inlining the functions so generics can be resolved at compile time.
Compilation failed
```

**Comment by alfonsogarciacaro** [comment, marked as answer]
There's some reflection support in Fable, but objects don't carry extra type info with them so it must be resolved at compile time (this is by design so Fable apps are compatible with JS tree shaking). The usual pattern is to offer two functions: one generic and inlined that gets the type info argument and just calls another function (not inlined) that receives the `Type` argument. You can see this pattern in Thoth.Json for example: https://github.com/thoth-org/Thoth.Json/blob/358ba12ac50ebac175bbfea98f9df20c0ddd1189/src/Decode.fs#L1233-L1238
